src/Lean/Meta/Tactic/Simp/Main.lean

```diff
@@ -402,41 +380,400 @@ def simpForall (e : Expr) : SimpM Result := withParent e do
+    match e with
+    | .letE n t v b true =>
+      let typeBackDeps := getBackDeps t
+      let valueBackDeps := getBackDeps v
+      let t := t.instantiateRev fvars
+      let v := v.instantiateRev fvars
+      let level ← withLCtx' lctx <| getLevel t
+      let fvarId ← mkFreshFVarId
+      let decl := LocalDecl.ldecl 0 fvarId n t v true .default
+      let info := { info with haveInfo := info.haveInfo.push { typeBackDeps, valueBackDeps, decl, level } }
+      let lctx := lctx.addDecl decl
+      let fvars := fvars.push (mkFVar fvarId)
+      collect b (numHaves + 1) info lctx fvars
+    | _ =>
+      let bodyDeps := getBackDeps e
+      withLCtx' lctx do
+        let body := e.instantiateRev fvars
+        let bodyType ← inferType body
+        let level ← getLevel bodyType
+        -- Collect fvars appearing in the type of `e`. Computing `bodyType` in particular is where `MetaM` is necessary.
+        let bodyTypeFVarIds := (collectFVars {} bodyType).fvarSet
+        let bodyTypeDeps : Std.HashSet Nat := Nat.fold fvars.size (init := {}) fun idx _ deps =>
+          if bodyTypeFVarIds.contains fvars[idx].fvarId! then
+            deps.insert idx
+          else
+            deps
+        return { info with bodyDeps, bodyTypeDeps, body, bodyType, level }
+
+/--
+Computes which `have`s in the telescope are fixed and which are unused.
+The length of the unused array may be less than the number of `have`s: use `unused.getD i true`.
+-/
+def HaveTelescopeInfo.computeFixedUsed (info : HaveTelescopeInfo) (keepUnused : Bool) :
+    MetaM (Std.HashSet Nat × Array Bool) := do
+  if keepUnused then
+    return (info.bodyTypeDeps, #[])
+  let numHaves := info.haveInfo.size
+  let updateArrayFromBackDeps (arr : Array Bool) (s : Std.HashSet Nat) : Array Bool :=
+    s.fold (init := arr) fun arr idx => arr.set! idx true
+  let mut used : Array Bool := Array.replicate numHaves false
+  -- Initialize `used` with the body's dependencies.
+  -- There is no need to consider `info.bodyTypeDeps` in this computation.
+  used := updateArrayFromBackDeps used info.bodyDeps
+  -- For each used `have`, in reverse order, update `used`.
+  for i in [0:numHaves] do
+    let idx := numHaves - i - 1
+    if used[idx]! then
+      let hinfo := info.haveInfo[idx]!
+      used := updateArrayFromBackDeps used hinfo.typeBackDeps
+      used := updateArrayFromBackDeps used hinfo.valueBackDeps
+  return (info.bodyTypeDeps, used)
+
+/--
+Auxiliary structure used to represent the return value of `simpHaveTelescopeAux`.
+-/
+private structure SimpHaveResult where
+  /--
+  The simplified expression in `(fun x => b) v` form. Note that it may contain loose bound variables.
+  `simpHaveTelescope` attempts to minimize the quadratic overhead imposed
+  by the locally nameless discipline in a sequence of `have` expressions.
+  -/
+  expr     : Expr
+  /--
+  The type of `expr`. It may contain loose bound variables like in the `expr` field.
+  Used in proof construction.
+  -/
```

Comment:
@kmill You say a few times it the doc strings in this structure that fields may contain loose bound variables, but you don't explain how it happens. I am assuming the "why" is "minimize the quadratic overhead imposed by the locally nameless approach".